Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabContent.tsx

```diff
@@ -0,0 +1,326 @@
+      if (watchedVar) {
+        // the variable is already being watched, so display a validation error under the picker
+        setPickerErrorMessages({
+          ...pickerErrorMessages,
+          [variableId]: formatMessage('You are already watching this property.'),
+        });
+      } else {
+        setWatchedVars({
+          ...watchedVars,
+          [variableId]: path,
+        });
+        // clear any error messages for the variable
+        delete pickerErrorMessages[variableId];
+        setPickerErrorMessages({ ...pickerErrorMessages });
+      }
+    },
+    [pickerErrorMessages, watchedVars]
+  );
+
+  // we need to refresh the details list every time a new bot state comes in
+  const refreshedWatchedVars = useMemo(() => {
+    return Object.entries(watchedVars).map(([key, value]) => {
+      return {
+        key,
+        value,
+      };
+    });
+  }, [mostRecentBotState, pickerErrorMessages, watchedVars]);
+
+  const renderRow = useCallback((props?: IDetailsRowProps) => {
+    if (props) {
+      return <DetailsRow {...props} styles={rowStyles} />;
+    }
+    return null;
+  }, []);
+
+  const renderColumn = useCallback(
+    (item: { key: string; value: string }, index: number | undefined, column: IColumn | undefined) => {
+      if (column && index !== undefined) {
+        if (column.key === NameColumnKey) {
+          // render picker
+          return (
+            <WatchVariablePicker
+              key={item.key}
+              errorMessage={pickerErrorMessages[item.key]}
+              path={item.value}
+              payload={memoryVariablesPayload}
+              variableId={item.key}
+              onSelectPath={onSelectPath}
+            />
+          );
+        } else if (column.key === ValueColumnKey) {
+          // render the value display
+          if (mostRecentBotState) {
+            const value = getValueFromBotTraceScope(item.value, mostRecentBotState?.activity);
+            if (value !== null && typeof value === 'object') {
+              // render monaco view
+              return <WatchTabObjectValue value={value} />;
+            } else if (value === undefined) {
+              // render undefined indicator
+              return <span css={undefinedValue}>undefined</span>;
+            } else {
+              // render primitive view
+              return <span>{String(value)}</span>;
+            }
+          } else {
+            // no bot trace available;
+            // render undefined indicator
+            return <span css={undefinedValue}>undefined</span>;
+          }
+        }
+      }
+      return null;
+    },
+    [pickerErrorMessages, mostRecentBotState, memoryVariablesPayload, watchedVars]
+  );
+
+  const onClickAdd = useCallback(() => {
+    setWatchedVars({
+      ...watchedVars,
+      [uuidv4()]: '',
+    });
+  }, [watchedVars]);
+
+  const onClickRemove = () => {
+    const updated = produce(watchedVars, (draftState) => {
```

Comment:
this is an overkill to bring in immer here, just delete and spread maybe instead?